### PR TITLE
Add WITH SESSION clause

### PR DESCRIPTION
**Pull request by wendigo**
Release notes: `Add WITH SESSION clause`

```
 WITH
    SESSION session_key = 'property',
    SESSION catalog.catalog_session_key = 'catalog_property'
SELECT 1
```

### CODE REVIEW

**Comment by wendigo** [comment]
@sajjoseph I was thinking exactly about these use-cases :)

**Comment by wendigo** [comment]
@martint PTAL

**Comment by wendigo** [comment]
@martint ping

**Comment by dain** [comment]
This is a cool idea

File: core/trino-main/src/main/java/io/trino/sql/SessionSpecificationInterpreter.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    public static ResolvedSessionSpecifications evaluate(Statement statement)
+    {
+        if (statement instanceof Query queryStatement) {
+            return evaluate(queryStatement.getSessionProperties());
+        }
+
+        return new ResolvedSessionSpecifications(ImmutableMap.of(), ImmutableMap.of());
+    }
+
+    private static ResolvedSessionSpecifications evaluate(List<SessionSpecification> specifications)
```
Comment (electrum) [inline core/trino-main/src/main/java/io/trino/sql/SessionSpecificationInterpreter.java:51]: Use `SessionPropertyManager.evaluatePropertyValue()` for this. Session properties support full expressions, not just literals. See `SetSessionTask` for how this should work.

**Comment by wendigo** [comment]
@electrum I've reworked this change. There is a chicken and egg problem here - session needs to have a transactionId in order to get a cataloghandle in order to validate passed catalog properties. Maybe there is a better way to apply this but I couldn't find any

File: core/trino-grammar/src/main/antlr4/io/trino/grammar/sql/SqlBase.g4 [hunk 2]
```diff
@@ -198,11 +202,23 @@ statement
...
 withFunction
-    : WITH functionSpecification (',' functionSpecification)*
+    : functionSpecification (',' functionSpecification)*
+    ;
+
+withSession
+    : sessionSpecification (',' sessionSpecification)*
+    ;
+
+sessionSpecification
```
Comment (martint) [inline core/trino-grammar/src/main/antlr4/io/trino/grammar/sql/SqlBase.g4:220]: The session is comprised of the set of property assignments, so a syntax that would be more appropriate is:

```
WITH SESSION
    property1 = value1,
    property2 = value2,
    ...
```

**Comment by losipiuk** [comment]
Can `WITH SESSION` be used as part as VIEW definition? This could be a workaround to settle things like hive timestamp precision, so we do not depend on default configuration of catalog.

File: core/trino-grammar/src/main/antlr4/io/trino/grammar/sql/SqlBase.g4 (hunk 2 above)
Comment (wendigo) [reply core/trino-grammar/src/main/antlr4/io/trino/grammar/sql/SqlBase.g4:220]: This change makes grammar ambigous as:

```
namedQuery
    : name=identifier (columnAliases)? AS '(' query ')'
    ;
with
    : WITH RECURSIVE? namedQuery (',' namedQuery)*
    ;
query
    : with? queryNoWith
    ;
rootQuery
    : queryScoped? query
    ;

queryScoped
    : WITH withFunction? withSession?
    ;
```

which in turn makes this SQL invalid:
```
WITH  SESSION session_key = 'property', catalog.catalog_session_key = 'catalog_property', catalog.catalog_session_key2 = true
SELECT foo()
```

fails with:
```
io.trino.sql.parser.ParsingException: line 2:13: mismatched input '='. Expecting: '(', 'AS'
	at io.trino.sql.parser.ErrorHandler.syntaxError(ErrorHandler.java:108)
	at org.antlr.v4.runtime.ProxyErrorListener.syntaxError(ProxyErrorListener.java:41)
	at org.antlr.v4.runtime.Parser.notifyErrorListeners(Parser.java:544)
	at org.antlr.v4.runtime.DefaultErrorStrategy.reportInputMismatch(DefaultErrorStrategy.java:327)
	at org.antlr.v4.runtime.DefaultErrorStrategy.reportError(DefaultErrorStrategy.java:139)
	at io.trino.grammar.sql.SqlBaseParser.namedQuery(SqlBaseParser.java:8679)
```

Comment (wendigo) [reply core/trino-grammar/src/main/antlr4/io/trino/grammar/sql/SqlBase.g4:220]: How can I make this non-ambigous @martint ?

**Pull request closed without merging**